### PR TITLE
[PR] Stage 1 - No `auth`, anonymous API for `Timers` and `Items`

### PR DESCRIPTION
**Pull request by LuchoTurtle**
related to #256 

This PR creates a first, basic, fully-tested version of a `RESTful` API for manipulating data anonymously with `timers` and `items`.
Also creates a document to detail its development.

### CODE REVIEW

File: lib/app_web/controllers/api/item_controller.ex [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  alias App.Item
+  import Ecto.Changeset
+
+  def show(conn, params) do
+    id = Map.get(params, "id")
+
+    try do
+      item = Item.get_item!(id)
+      json(conn, item)
+    rescue
```
[the remark this answers is not in the document]
Comment (LuchoTurtle) [reply lib/app_web/controllers/api/item_controller.ex]: Fixed in [`0abf8af` (#263)](https://github.com/dwyl/mvp/pull/263/commits/0abf8af3f95b341fa5e04ff4c93e7259e322de7c).
Thanks! 👍

**Comment by nelsonic** [comment]
https://github.com/dwyl/mvp/actions/runs/3906526566/jobs/6674798916#step:8:127
image


```sh
  1) test timers update_timer(%{id: id, start: start, stop: stop}) with start later than stop 
should throw error (App.TimerTest)
Error:      test/app/timer_test.exs:100
     ** (KeyError) key :id not found in: 
%{model: %App.Item{__meta__: #Ecto.Schema.Metadata<:loaded, "items">, 
id: 4, person_id: 1, status: nil, text: "some text", 
timer: #Ecto.Association.NotLoaded<association :timer is not loaded>, 
tags: #Ecto.Association.NotLoaded<association :tags is not loaded>, 
inserted_at: ~N[2023-01-12 22:24:06], updated_at: ~N[2023-01-12 22:24:06]}, 
version: %PaperTrail.Version{__meta__: #Ecto.Schema.Metadata<:loaded, "versions">, id: 4, 
event: "insert", item_type: "Item", item_id: 4, 
item_changes: %{id: 4, inserted_at: ~N[2023-01-12 22:24:06], person_id: 1, 
status: nil, text: "some text", updated_at: ~N[2023-01-12 22:24:06]}, 
originator_id: 1, origin: nil, meta: nil, inserted_at: 
~U[2023-01-12 22:24:06Z]}}
Warning:      code: Timer.start(%{item_id: item.id, person_id: 1, start: seven_seconds_ago})
     stacktrace:
       test/app/timer_test.exs:111: (test)
```

**Comment by nelsonic** [comment]
@LuchoTurtle tests failing due to me merging `main` branch into this one after merging your `PaperTrail` PR #255 💭 
Feel free to revert changes if you think it will fix the tests. Or we can pair on fixing the tests tomorrow morning. 🧑‍💻

**Comment by nelsonic** [comment]
@LuchoTurtle https://github.com/martinthenth/goal looks decent. 👍 
Please consider opening a separate issue for it: https://github.com/dwyl/technology-stack/issues 🆕 
I'm more inclined to keep this as dependency-free as possible for now. 🔰

**Comment by LuchoTurtle** [comment]
This should be done. 
The coverage "dropped" because of the line we all know it's covered but `excoveralls` skips, for some reason 😭

**Comment by nelsonic** [comment]
@LuchoTurtle you've assigned to me but still **`in-progress`**? 👨‍💻 
Please LMK if you feel it's **`awaiting-review`**. 👌

**Comment by LuchoTurtle** [comment]
fixed

**Comment by LuchoTurtle** [comment]
@nelsonic 
found a bug. So I'm going to retract your assignment status.

File: api.md [hunk 2]
```diff
@@ -0,0 +1,779 @@
...
+- `test/app_web/api/timer_controller_test.exs`
+
+Before implementing,
+we recommend giving a look at [`learn-api-design`](https://github.com/dwyl/learn-api-design),
+we are going to be using some tips coming from there!
+
+We want the API requests to be handled gracefully 
+when an error occurs.
+The person using the API 
+[**should be shown meaningful errors**](https://github.com/dwyl/learn-api-design/blob/revamp/README.md#show-meaningful-errors).
```
Comment (nelsonic) [inline api.md:95]: Let's get https://github.com/dwyl/learn-api-design/pull/47 reviewed/updated & merged _now_ so that we can update this link. 🔗

Comment (nelsonic) [reply api.md:95]: https://github.com/dwyl/learn-api-design/pull/47 merged. ✅

**Comment by nelsonic** [comment]
`AppWeb.API.TimerControllerTest` -> `API.TimerTest` 😉

**Comment by nelsonic** [comment]
Verbose instructions:

```md
Let's head over and create our `TimerController`!
Inside the same directory, create `timer_controller.ex`
and use this code.
```
Replaced with:
```md
Create a new file with the path:
`lib/api/timer.ex`
and add the following code:
```

Which do you think is easier for the _reader_? 💭

**Comment by nelsonic** [comment]
```
 To github.com:dwyl/app-mvp.git
 ! [rejected]        api-#256 -> api-#256 (non-fast-forward)
error: failed to push some refs to 'github.com:dwyl/app-mvp.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
n@mbp16 mvp % git pull
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint:
hint:   git config pull.rebase false  # merge
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint:
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.
```
# 😢 

PR assigned to me but updated and now my review updates are rejected. 😢 
Such an epic waste of my extremely limited time. ⏳ 💸 🔥

**Comment by LuchoTurtle** [comment]
The PR is assigned to you and I haven't pushed any of my changes because I'm waiting for you to assign it to me so I can update the bug I found earlier. 

I don't know how you got this issue because this PR hasn't been updated by anyone else ever since  👀🤔

**Comment by nelsonic** [comment]
@LuchoTurtle please open a `new` issue for the bug you spotted. 🐛 🆕  🙏  (thanks!)

**Comment by LuchoTurtle** [comment]
The bug is in the implementation in this PR, I just need to make a couple of changes to it. No reason to merge this until I get it fixed 😅

**Comment by LuchoTurtle** [comment]
Found it after assigning it for review, my fault on that. But it's if quick resolution, just assign it back to me when it's convenient for you

**Comment by nelsonic** [comment]
Cool. I've done a review and made a few minor updates. 🧑‍💻 
Assigning back to you _now_ for the update. 👌 
You shouldn't have any more issues with CodeCov complaining about 1 line not being covered. 🤞 
Assign back once you're done. ✅

**Comment by LuchoTurtle** [comment]
The merge should have fixed the problem.
Thanks!

**Comment by LuchoTurtle** [comment]
> `AppWeb.API.TimerControllerTest` -> `API.TimerTest` 😉

Just for clarification, I followed a "convention" that appeared on Stack Overflow -> https://stackoverflow.com/questions/42598281/where-to-put-phoenix-controllers-for-app-with-both-web-frontend-and-an-api

But I do agree your change makes more sense 👍

**Review by nelsonic — APPROVED** [review]
@LuchoTurtle Looks good. Thanks. 👍 
Made a couple of notes in: https://github.com/dwyl/mvp/issues/256#issuecomment-1384091996 and opened a couple of new follow-up / side-quest issues: #269 & #268 

Merging. :shipit:

**Comment by nelsonic** [comment]
Testing the `/api` of the Review App:
```sh
curl -X POST https://mvp-pr-263.fly.dev/api/items -H 'Content-Type: application/json' -d '{"text":"My Awesome item text"}' 
```

Response:

```sh
{"id":1}
```

Satisfied this is working. ✅

**Pull request merged**